src/dictionary.c

```diff
@@ -160,6 +158,37 @@ void dict_put(struct dictionary* d, uint32_t hash, R_len_t i) {
   d->used++;
 }
 
+// Assume worst case, that every value is distinct, aiming for a load factor
+// of at most 77%. We round up to power of 2 to ensure quadratic probing
+// strategy works. Maximum power of 2 we can store in a uint32_t is 2^31,
+// as 2^32 is 1 greater than the max uint32_t value, so we clamp sizes that
+// would result in 2^32 to INT_MAX to ensure that our maximum ceiling value
+// is only 2^31. This will increase the load factor above 77% for `x` with
+// length greater than 1653562409 (2147483648 * .77), but it ensures that
+// it can run.
+static inline
+uint32_t dict_key_size(SEXP x) {
+  const R_len_t x_size = vec_size(x);
+
+  if (x_size > R_LEN_T_MAX) {
+    stop_internal("dict_key_size", "Dictionary functions do not support long vectors.");
+  }
+
+  const double load_adjusted_size = x_size / 0.77;
+
+  if (load_adjusted_size > UINT32_MAX) {
+    stop_internal("dict_key_size", "Can't safely cast load adjusted size to a `uint32_t`.");
+  }
+
+  uint32_t size = (uint32_t)load_adjusted_size;
+  size = size > INT_MAX ? INT_MAX : size;
```

Comment:
As mentioned on Slack we should also check that the final `size` is larger than the length of the vector to be defensive. Open addressing doesn't support load factors > 100%.

Reply:
The point of the clamping to `INT_MAX` is to ensure that we can't exceed a load factor of 100%.

This will all have to be updated when we change to long vectors anyways, and the `x_size > R_LEN_T_MAX` check will catch the long vectors before they cause any issues here.

Would a comment saying that this will need to be adjusted for long vectors suffice?

Reply:
I went ahead and added the check e560cdf661b997be51e50294d0d0f910a63a9b52

Reply:
> The point of the clamping to INT_MAX is to ensure that we can't exceed a load factor of 100%.

Reducing the adjusted size will only increase the worst case load factor. Aren't we clamping to INT_MAX to avoid `ceil2()` overflow? It would be useful to have an inline comment about this here as well.

```c
if (x_size > size) { r_stop_internal("key size must be at least as large as x"); }
```

That sounds good thanks! Though I'd use "hash table size" rather than "key size".

```c
if (x_size > size) {
  r_stop_internal(, "Hash table size must be at least as large as input.");
}
```